packages/core/content-manager/server/src/controllers/relations.ts

```diff
@@ -45,117 +43,196 @@ const sanitizeMainField = (model: any, mainField: any, userAbility: any) => {
   return mainField;
 };
 
+const mapResults = (results: Array<any>) => {
+  return results.map((result: any) => {
+    if (result.documentId !== undefined) {
+      result.id = result.documentId;
+      delete result.documentId;
+    }
+
+    return result;
+  });
+};
+
 export default {
-  async findAvailable(ctx: any) {
+  async extractAndValidateRequestInfo(
+    ctx: any,
+    id?: Entity.ID,
+    status?: Documents.Params.PublicationState.Kind,
+    locale?: Documents.Params.Attribute.Locale
+  ) {
     const { userAbility } = ctx.state;
     const { model, targetField } = ctx.params;
 
-    await validateFindAvailable(ctx.request.query);
-
-    // idsToOmit: used to exclude relations that the front already added but that were not saved yet
-    // idsToInclude: used to include relations that the front removed but not saved yes
-    const { entityId, idsToOmit, idsToInclude, _q, ...query } = ctx.request.query;
-
-    const modelSchema = strapi.getModel(model);
-    if (!modelSchema) {
-      return ctx.badRequest("The model doesn't exist");
+    const sourceSchema = strapi.getModel(model);
+    if (!sourceSchema) {
+      throw new errors.ValidationError(`The model ${model} doesn't exist`);
     }
 
-    const attribute: any = modelSchema.attributes[targetField];
+    const attribute: any = sourceSchema.attributes[targetField];
     if (!attribute || attribute.type !== 'relation') {
-      return ctx.badRequest("This relational field doesn't exist");
+      throw new errors.ValidationError(
+        `The relational field ${targetField} doesn't exist on ${model}`
+      );
     }
 
-    const isComponent = modelSchema.modelType === 'component';
-
-    if (!isComponent) {
-      const permissionChecker = getService('permission-checker').create({
-        userAbility,
-        model,
-      });
+    const permissionChecker = getService('permission-checker').create({
+      userAbility,
+      model,
+    });
 
+    const isSourceComponent = sourceSchema.modelType === 'component';
+    if (!isSourceComponent) {
       if (permissionChecker.cannot.read(null, targetField)) {
         return ctx.forbidden();
       }
+    }
 
-      if (entityId) {
-        const entityManager = getService('entity-manager');
+    const where: Record<string, any> = {};
+    if (!isSourceComponent) {
+      where.locale = locale;
+      where.publishedAt = status === 'published' ? { $ne: null } : null;
```

Comment:
mental note, create a function in strapi utils to transform status to published at filter , I added this exact same line many times in the doc service